From 79f59eb6430551cf067bc84da2b462a6cc6de5eb Mon Sep 17 00:00:00 2001
From: Xiao Chen <xiao@apache.org>
Date: Thu, 15 Jun 2017 10:40:59 -0700
Subject: [PATCH 2521/2748] HADOOP-14523. OpensslAesCtrCryptoCodec.finalize()
 holds excessive amounts of memory. Contributed by
 Misha Dmitriev.

(cherry picked from commit ef8edab930338646551cbe3c7e7cf954e21c0f9a)
(cherry picked from commit 7d81b0beab3de7d66c6200e925bb7ac991772728)

Change-Id: If438f17cf1a01269393e9deda7fda7aa6a38d85f
---
 .../hadoop/crypto/OpensslAesCtrCryptoCodec.java    |   11 -----------
 .../hadoop/crypto/random/OsSecureRandom.java       |    5 +++++
 2 files changed, 5 insertions(+), 11 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/OpensslAesCtrCryptoCodec.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/OpensslAesCtrCryptoCodec.java
index 63c382c..d0a12e9 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/OpensslAesCtrCryptoCodec.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/OpensslAesCtrCryptoCodec.java
@@ -19,7 +19,6 @@
 
 import static org.apache.hadoop.fs.CommonConfigurationKeysPublic.HADOOP_SECURITY_SECURE_RANDOM_IMPL_KEY;
 
-import java.io.Closeable;
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.security.GeneralSecurityException;
@@ -72,16 +71,6 @@ public void setConf(Configuration conf) {
   }
 
   @Override
-  protected void finalize() throws Throwable {
-    try {
-      Closeable r = (Closeable) this.random;
-      r.close();
-    } catch (ClassCastException e) {
-    }
-    super.finalize();
-  }
-
-  @Override
   public Configuration getConf() {
     return conf;
   }
diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/random/OsSecureRandom.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/random/OsSecureRandom.java
index 2010e6d..9428b98 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/random/OsSecureRandom.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/random/OsSecureRandom.java
@@ -116,4 +116,9 @@ synchronized public void close() {
       stream = null;
     }
   }
+
+  @Override
+  protected void finalize() throws Throwable {
+    close();
+  }
 }
-- 
1.7.9.5

